src/OpenCL.cpp

```diff
@@ -148,11 +156,12 @@ __kernel void out_transform(__global float *M, __global net_t *Y,
     int y = 2*block_y;
 
     if (k < K && block < P) {
+        int kHW = k * HW;
         int b = block_y * WTILES + block_x;
         float temp_m[16];
-        for (int xi = 0; xi < 4; xi++) {
+        for (int xi = 0; xi < 16; xi+=4) {
             for (int nu = 0; nu < 4; nu++) {
-                temp_m[xi*4 + nu] = M[xi*(4*Kpad*Ppad) + nu*(Kpad*Ppad)+ b*Kpad + k];
+                temp_m[xi + nu] = M[(xi + nu)*(KPpad)+ b*Kpad + k];
```

Comment:
b*Kpad might be worth extracting as it doesn't change, same for the loop below. Another thing is extracting xi * KPpad to the outer loop, and just adding KPpad each inner loop to replace nu*KPpad with simple addition, up to you.

Awesome changes by the way, nice to optimize away expensive operations.